File: app/src/main/kt/estyle/teabaike/activity/ContentActivity.java
package estyle.teabaike.activity;

import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.support.design.widget.Snackbar;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;

import java.lang.reflect.Method;

import javax.inject.Inject;

import estyle.teabaike.R;
import estyle.teabaike.application.TeaBaikeApplication;
import estyle.teabaike.bean.ContentDataBean;
import estyle.teabaike.databinding.ActivityContentBinding;
import estyle.teabaike.manager.GreenDaoManager;
import estyle.teabaike.manager.RetrofitManager;
import io.reactivex.disposables.Disposable;

public class ContentActivity extends BaseActivity {

    private Snackbar mSnackbar;

    private ActivityContentBinding binding;

    private ContentDataBean mData;

    @Inject
    RetrofitManager mNetworkProvider;
    @Inject
    GreenDaoManager mDBProvider;

    private Disposable mDisposable;

    public static void startActivity(Context context, long id, boolean isOnline) {
        Intent intent = new Intent(context, ContentActivity.class);
        intent.putExtra("id", id);
        intent.putExtra("is_online", isOnline);
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        TeaBaikeApplication.getInstance().getTeaBaikeComponent().inject(this);
        binding = DataBindingUtil.setContentView(this, R.layout.activity_content);

        initView();
        initData();
    }

    private void initView() {
        setSupportActionBar(binding.toolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
    }

    private void initData() {
        long id = getIntent().getLongExtra("id", 0);
        boolean isOnline = getIntent().getBooleanExtra("is_online", false);
        if (isOnline) {
            mDisposable = mNetworkProvider.loadContentData(id)
                    .subscribe(contentDataBean -> {
                                mData = contentDataBean;
                                binding.setBean(mData);
                            },
                            throwable -> Toast.makeText(ContentActivity.this,
                                    R.string.fail_connect,
                                    Toast.LENGTH_SHORT).show());
        } else {
            mData = mDBProvider.queryCollectionDataById(id);
            binding.setBean(mData);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_content, menu);

        Class clazz = menu.getClass();
        Method setOptionalIconsVisibleMethod;
        try {
            setOptionalIconsVisibleMethod = clazz.getDeclaredMethod("setOptionalIconsVisible"
                    , boolean.class);
            setOptionalIconsVisibleMethod.setAccessible(true);
            setOptionalIconsVisibleMethod.invoke(menu, true);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.share_item:
                share();
                break;
            case R.id.collect_item:
                collect();
                break;
            case android.R.id.home:
                finish();
                break;
        }
        return super.onOptionsItemSelected(item);
    }

    // 分享文章
    private void share() {
        if (mData != null) {
            showTip(R.string.share_successful);
        }
    }

    // 收藏文章
    private void collect() {
        if (mData != null) {
            mDBProvider.collectData(mData);
            showTip(R.string.collect_successful);
        }
    }

    // 展示Snackbar
    private void showTip(int resId) {
        if (mSnackbar == null) {
            mSnackbar = Snackbar.make(binding.getRoot(), resId, Snackbar.LENGTH_SHORT);
            mSnackbar.getView().setBackgroundResource(R.color.colorAccent);
        } else {
            mSnackbar.setText(resId);
        }
        mSnackbar.show();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mDisposable != null) {
            mDisposable.dispose();
        }

        binding.contentWebView.removeAllViews();
        binding.contentWebView.destroy();
    }
}

File: app/src/main/kt/estyle/teabaike/activity/ContentActivity.kt
package estyle.teabaike.activity

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProviders
import com.uber.autodispose.ObservableSubscribeProxy
import estyle.base.BaseActivity
import estyle.base.rxjava.DisposableConverter
import estyle.base.rxjava.RefreshObserver
import estyle.base.rxjava.SnackbarObserver
import estyle.base.widget.Snackbar
import estyle.teabaike.R
import estyle.teabaike.databinding.ActivityContentBinding
import estyle.teabaike.entity.ContentEntity
import estyle.teabaike.viewmodel.ContentViewModel
import kotlinx.android.synthetic.main.activity_content.*
import java.lang.reflect.Method

class ContentActivity : BaseActivity() {

    private val id by lazy { intent.getLongExtra("id", 0) }

    // TODO 此处使用val不可以懒加载
    // 但是在HeadlineHeaderView中就可以
    private lateinit var binding: ActivityContentBinding
    private val viewModel by lazy { ViewModelProviders.of(this)[ContentViewModel::class.java] }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_content)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        swipe_refresh_layout.setOnRefreshListener { refresh() }
    }

    override fun onPostCreate(savedInstanceState: Bundle?) {
        super.onPostCreate(savedInstanceState)

        swipe_refresh_layout.post {
            swipe_refresh_layout.isRefreshing = true
            refresh()
        }
    }

    private fun refresh() {
        viewModel.refresh(id)
            .`as`<ObservableSubscribeProxy<ContentEntity.DataEntity>>(
                DisposableConverter.dispose(this)
            )
            .subscribe(object : RefreshObserver<ContentEntity.DataEntity>(swipe_refresh_layout) {
                override fun onNext(it: ContentEntity.DataEntity) {
                    super.onNext(it)
                    binding.content = it
                }
            })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_content, menu)
        val setOptionalIconsVisibleMethod: Method = menu.javaClass.getDeclaredMethod(
            "setOptionalIconsVisible",
            Boolean::class.javaPrimitiveType
        )
        setOptionalIconsVisibleMethod.isAccessible = true
        setOptionalIconsVisibleMethod.invoke(menu, true)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.share_item -> share()
            R.id.collect_item -> collect()
            android.R.id.home -> finish()
        }
        return super.onOptionsItemSelected(item)
    }

    // 收藏文章
    private fun collect() {
        viewModel.collect(binding.content!!)
            .`as`<ObservableSubscribeProxy<Long>>(DisposableConverter.dispose(this))
            .subscribe(object :
                SnackbarObserver<Long>(binding.root, getString(R.string.collect_failure)) {
                override fun onComplete() {
                    super.onComplete()
                    showTip(getString(R.string.collect_successful))
                }
            })
    }

    // 分享文章
    private fun share() {
        showTip(getString(R.string.share_successful))
    }

    // 展示Snackbar
    private fun showTip(message: String) {
        Snackbar.make(binding.root, message, Snackbar.LENGTH_SHORT)
            .show()
    }

    override fun onDestroy() {
        super.onDestroy()
        content_web_view.removeAllViews()
        content_web_view.destroy()
    }

    companion object {

        fun startActivity(context: Context, id: Long) {
            val intent = Intent(context, ContentActivity::class.java)
            intent.putExtra("id", id)
            context.startActivity(intent)
        }
    }
}
